### PR TITLE
Keyboard: Add sticky keys functionality

### PR DESCRIPTION
**Pull request by sypl**
Adding an option to reproduce sticky keys accessibility feature [https://en.wikipedia.org/wiki/Sticky_keys]

Tapping modifier keys in any sequence will add them to a register and apply them all with a keycode when next non-modifier keycode is sent through.

A second tap to a modifier will lock it, such that even after the keycode is sent with applied modifiers, locked modifiers will remain on the register.

Tapping shift five times toggles the feature on and off.

### CODE REVIEW

**Comment by drashna** [comment]
One Shot Mods does this already, I think: https://docs.qmk.fm/#/feature_advanced_keycodes?id=one-shot-keys

**Comment by sypl** [comment]
It's not possible to achieve this functionality with one shot mods, as far as I know. Mods clear after the post-mod key press. Additionally, there is a time limit as to how long after the mod is tapped that the feature applies.

No such restrictions apply here. Users can take as long as they need and apply up to all for mods. It's like a super mod one shot.

This is a standard accessibility feature that typically has needed OS software to enable. I thought being able to carry it on your own keyboard would be really useful.

**Comment by drashna** [comment]
For OSM, mods only clear if `ONESHOT_TIMEOUT` is defined.  If it's not defined, then they stay active until an appropriate keypress. 

And if you het another mod key, it will stack them, as well.

**Comment by drashna** [comment]
ANd I'm starting to see what you're trying to do here.  

👍

File: docs/feature_sticky_keys.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+> Sticky keys is an accessibility feature of some graphical user interfaces to assist users who have physical disabilities or help users reduce repetitive strain injury (or a syndrome called the Emacs Pinky). It serializes keystrokes instead of pressing multiple keys at a time, allowing the user to press and release a modifier key, such as Shift, Ctrl, Alt, or the Windows key, and have it remain active until any other key is pressed.
+
+## Usage
+
+First, enable Sticky Keys by setting `STICKY_KEYS_ENABLE = yes` in your `rules.mk`.
+
+Tap (either) `SHIFT` five times to toggle it on or off.
+
+Once toggled on, tap modifiers to have them apply to your next key, i.e. `SHIFT`, `CTRL`, `ALT`, `C` will send `SHIFT` + `CTRL` + `ALT` + `X`.
```
Comment (noroadsleft) [inline docs/feature_sticky_keys.md]: You have `C` on one side of the explanation and `X` on the other side, which if I understand this function correctly, would be a typo.

Comment (drashna) [reply docs/feature_sticky_keys.md]: I had assumed it was a typo, yeah.   Was more concerned about functionality, and assumed it was an offset error.

Comment (sypl) [reply docs/feature_sticky_keys.md]: Yep, typo, will fix.

**Comment by sypl** [comment]
@drashna I didn't know that about the `ONESHOT_TIMEOUT`. Theoretically it gets us about 80% there to this functionality. Can't really lock mods though, at least not deliberately.

I tried it just now by placing the following keys on a clean keyboard:

```
#define SK_LALT OSM(MOD_LALT)
#define SK_LCTL OSM(MOD_LCTL)
#define SK_LSFT OSM(MOD_LSFT)
#define SK_LGUI OSM(MOD_LGUI)
#define SK_RALT OSM(MOD_RALT)
#define SK_RCTL OSM(MOD_RCTL)
#define SK_RSFT OSM(MOD_RSFT)
#define SK_RGUI OSM(MOD_RGUI)
```

The sticky mods behaviour seems OK, though there might a slight bug. Some mods do lock if you mash around the mods and then hit another key. I'm not sure how to reproduce this consistently yet, but it does happen.

**Comment by drashna** [comment]
Yeah, I think you can see my concern about overlapping features.  

However, this does look distinct enough to me, that and it's definitely very neat. :)

**Comment by sypl** [comment]
For the OSM bug I mentioned just now, I'm wondering if it's related to the issue I raised previously: https://github.com/qmk/qmk_firmware/issues/3838

It certainly seems to be a similar behaviour, a layer/mod getting stick when I elide over two OSL/OSM keys at the same time. The OSM one seems more difficult to reproduce, but it happens often enough that it seems a real issue. I think I may re-open that issue and even look in to it.

**Comment by sypl** [comment]
I think I can close this PR. It seems you can achieve a very similar functionality just by putting this in a keymap's `config.h`:

```
#define PREVENT_STUCK_MODIFIERS
#define ONESHOT_TAP_TOGGLE 2
#undef ONESHOT_TIMEOUT
```

It works pretty darn well. If I was 80% towards sticky keys functionality with my previous attempt at using OSM only, I'd say i'm up to 98% now. 

I suspect there's still a subtle bug here somewhere. Mashing around on modifiers again I got in to a state where a modifier got stuck, with seemingly no way to clear it. I'll file an issue if I can reliably recreate it.  

Some things are slightly different than how macOS implements sticky keys (can't talk about other OS, haven't really tried sticky keys on them), whereby tapping twice on a mod, no matter the the mod sequence, will toggle it. So `SHIFT`, `ALT`, `SHIFT`, `GUI` would lock `SHIFT`.

In contrast the setup with `OSM` with `ONESHOT_TAP_TOGGLE` set to 2 means that you have to hit the same modifier twice in succession. I'd actually argue that this is a more intuitive method for toggling a layer.

I'm glad there's a way of achieving this functionality without expanding the core. I wonder though whether there need exist a place for 'recipes' like this. There's a lot of cool functionality that people are creating with this firmware, but you have do dig through the keymaps in source to find them.

**Pull request closed without merging**